### PR TITLE
Prevent unnecessary vectorization in default TileAndFuse Config Selector 

### LINKED ISSUE #23655: Small Workgroup sizes negatively affecting performance for Batch Norm Kernels
This issue is a follow-up triage of the previous issues [#21824](https://github.com/iree-org/iree/issues/21824#issuecomment-3955467083) & related to [#23539](https://github.com/iree-org/iree/issues/23539).

Batch Norm produces a total of 5 dispatches, 2 of which perform reduction. On these dispatches, the workgroup size set by the heuristic for different Batch-Norm shapes when using TileAndFuse pipeline is equal to the size of the subgroup. 
Performance improves when using a larger workgroup size with smaller thread tile. 

### Original Configuration
```
{translation_info = ...workgroup_size = [64, 1, 1] subgroup_size = 64>}
...
llinalg.generic ... attrs = #iree_gpu.lowering_config<{reduction = [0, 4], thread = [4, 0], workgroup = [256, 0]}>}
```
mean execution time = 176.26us
[Link](https://gist.github.com/YashDeshpande25/8ce64bdc914fdc3ca1a90106e95d7ed1.js) to the complete unmodified IR 

### Manually Edited Configuration
```
{translation_info = ...workgroup_size = [256, 1, 1] subgroup_size = 64>}
...
llinalg.generic ... attrs = #iree_gpu.lowering_config<{reduction = [0, 4], thread = [1, 0], workgroup = [256, 0]}>}
```
mean execution time = 102.67us
[Link](https://gist.github.com/YashDeshpande25/6ac75637618378506e62377154a76a59) to the manually updated IR 

Timed using the boo-driver. 

## Steps to reproduce 

Running batchnorm config 12 via the boo-driver:
```
iree-boo-driver bnormbfp16 --batchsize 16 --in_channels 2048 --in_h 48 --in_w 32 --time 1 --iter 100 --save 1 --run 1 --layout NHWC
```

To compile the generated file 

- Use the flag `--compile-to=executable-configurations` to generate the editable IR. 

After manually modifying the IR, to generate an updated vmfb file from the updated mlir file

- Use the flag `--compile-from=executable-configurations` 

Use the flag `--iree-codegen-llvmgpu-use-reduction-vector-distribution=false` to apply the TileAndFuse pipeline

### PR DESCRIPTION
**Pull request by YashDeshpande25**
This change updates the default TileAndFuse configuration selector to initially check for a workgroup size of 256 before resorting to the standard workgroup size of  64 which increases GPU resource utilization. It also adds an extra check to prevent vectorization when it hinders performance in cases where each thread has considerable reduction work. 

Fixes [#23655](https://github.com/iree-org/iree/issues/23655)

### CODE REVIEW

File: compiler/src/iree/compiler/Codegen/Dialect/GPU/TargetUtils/ConfigUtils.cpp [hunk 1]
```diff
@@ -1412,6 +1413,65 @@ static FailureOr<DistributionInfo> collectOpDistributionInfo(Operation *op) {
   return distInfo;
 };
 
+/// Check if the reduction op has a single combiner operation.
+static bool hasSingleReductionCombiner(linalg::LinalgOp op) {
```
Comment (nirvedhmeshram) [inline compiler/src/iree/compiler/Codegen/Dialect/GPU/TargetUtils/ConfigUtils.cpp]: I currently dont follow why this is needed but maybe once I see some tests, with and without single redcution combiner I will get it.

**Review by nirvedhmeshram — COMMENTED** [review]
Idea makes sense.

File: compiler/src/iree/compiler/Codegen/Dialect/GPU/TargetUtils/ConfigUtils.cpp [hunk 2]
```diff
@@ -1615,7 +1682,8 @@ LogicalResult setTileAndFuseLoweringConfig(IREE::GPU::TargetAttr target,
   };
 
   // First try to see if we can use up all threads without any loss.
-  int64_t newNumThreads = subgroupSize;
+  // Check with 256 as the preferred size of workgroups for AMD GPUs.
```
Comment (nirvedhmeshram) [inline compiler/src/iree/compiler/Codegen/Dialect/GPU/TargetUtils/ConfigUtils.cpp]: I would just rephrase to say try with at least four sub groups first per workgroup to have good occupancy and hardware utilization.

File: compiler/src/iree/compiler/Codegen/Dialect/GPU/TargetUtils/ConfigUtils.cpp (hunk 1 above)
Comment (YashDeshpande25) [reply compiler/src/iree/compiler/Codegen/Dialect/GPU/TargetUtils/ConfigUtils.cpp]: Sure, I added it to match reductionConfigUtils.cpp I will look into the tests that they have added for it to see if any are relevant to the current requirement.

**Review by Groverkss — COMMENTED** [review]
I think this would need more context

**Comment by YashDeshpande25** [comment]
Batch Norm kernels produce 2 reduction dispatches which use the channel dim as their parallel dimension & carry out reduction across the rest of the dimensions. 
I noticed that in the TileAndFuse pipeline we were losing performance due to default workgroup size of 64 & vectorization set to 4. 

## For shape-12

```
iree-boo-driver bnormbfp16 --batchsize 16 --in_channels 2048 --in_h 48 --in_w 32 --time 1 --iter 100 --save 1 --run 1 --layout NHWC
```

### Original Configuration
```
{translation_info = ...workgroup_size = [64, 1, 1] subgroup_size = 64>}
...
llinalg.generic ... attrs = #iree_gpu.lowering_config<{reduction = [0, 4], thread = [4, 0], workgroup = [256, 0]}>}
```
mean execution time = 176.26us

But that could be improved upon if we avoided any default assumed vectorization & increased the workgroup sizes to fit the hardware. 

### Updated Configuration
```
{translation_info = ...workgroup_size = [256, 1, 1] subgroup_size = 64>}
...
llinalg.generic ... attrs = #iree_gpu.lowering_config<{reduction = [0, 4], thread = [1, 0], workgroup = [256, 0]}>}
```
mean execution time = 102.67us


## For shape-11

```
iree-boo-driver bnormbfp16 --batchsize 16 --in_channels 2048 --in_h 16 --in_w 32 --time 1 --iter 100 --save 1 --run 1 --layout NHWC
```

Original Execution Time: 116.48us
Updated Execution Time: 71.04us


Based on the current default TAF config selector this should benefit the performance of shapes 11 & 12 both of which have large channel dimensions. 

I have added some relevant tests to demonstrate the functioning.
